src/wmtk/EdgeMeshOperationExecutor.cpp

```diff
@@ -0,0 +1,238 @@
+    assert(new_vids.size() == 1);
+    const long v_new = new_vids[0];
+    // create new edges
+    // new_eids[i] is connect to m_neighbor_eids[i] and m_spine_vids[i]
+    const std::vector<long> new_eids = this->request_simplex_indices(PrimitiveType::Edge, 2);
+    assert(new_eids.size() == 2);
+
+    const long local_vid = m_mesh.is_ccw(m_operating_tuple) ? 0 : 1;
+
+    // update ee
+    {
+        // for 2 new edges
+        auto ee_new_0 = ee_accessor.index_access().vector_attribute(new_eids[0]);
+        auto ee_new_1 = ee_accessor.index_access().vector_attribute(new_eids[1]);
+        ee_new_0[(local_vid + 1) % 2] = new_eids[1];
+        ee_new_1[local_vid] = new_eids[0];
+        if (m_is_self_loop) {
+            ee_new_0[local_vid] = new_eids[1];
+            ee_new_1[(local_vid + 1) % 2] = new_eids[0];
+        } else {
+            ee_new_0[local_vid] = m_neighbor_eids[0];
+            ee_new_1[(local_vid + 1) % 2] = m_neighbor_eids[1];
+            // for neighbor edges
+            for (long i = 0; i < 2; i++) {
+                if (m_neighbor_eids[i] != -1) {
+                    auto ee_neighbor =
+                        ee_accessor.index_access().vector_attribute(m_neighbor_eids[i]);
+                    auto ev_neighbor =
+                        ev_accessor.index_access().vector_attribute(m_neighbor_eids[i]);
+                    for (long j = 0; j < 2; j++) {
+                        if (ee_neighbor[j] == m_operating_edge_id && ev_neighbor[j] == m_spine_vids[i]) {
+                            ee_neighbor[j] = new_eids[i];
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // update ev
+    {
+        // for new edges
+        auto ev_new_0 = ev_accessor.index_access().vector_attribute(new_eids[0]);
+        auto ev_new_1 = ev_accessor.index_access().vector_attribute(new_eids[1]);
+        ev_new_0[local_vid] = m_spine_vids[0];
+        ev_new_0[(local_vid + 1) % 2] = v_new;
+        ev_new_1[local_vid] = v_new;
+        ev_new_1[(local_vid + 1) % 2] = m_spine_vids[1];
+    }
+
+    // update ve
+    {
+        // for new vertex
+        ve_accessor.index_access().scalar_attribute(v_new) = new_eids[0];
+
+        // for spine vertices
+        ve_accessor.index_access().scalar_attribute(m_spine_vids[0]) = new_eids[0];
+        ve_accessor.index_access().scalar_attribute(m_spine_vids[1]) = new_eids[1];
+    }
+    update_cell_hash();
+    delete_simplices();
+
+    // prepare return Tuple
+    return m_mesh.edge_tuple_from_id(new_eids[0]);
+}
+
+void EdgeMesh::EdgeMeshOperationExecutor::update_hash_in_map(EdgeMesh& child_mesh)
+{
+    // TODO: Implement for multi_mesh in the future
+}
+
+Tuple EdgeMesh::EdgeMeshOperationExecutor::collapse_edge()
+{
+    return collapse_edge_single_mesh();
+    // TODO: Implement for multi_mesh in the future
```

Comment:
If the mesh is part of a multi-mesh, throw an exception.

Reply:
This function is already used in tests.